texar/modules/pretrained/xlnet_model_utils.py

```diff
@@ -0,0 +1,342 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Model Utils of XLNet Modules.
+
+Adapted from
+https://github.com/zihangdai/xlnet/blob/master/modeling.py
+"""
+
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from texar.core import get_layer
+from texar.module_base import ModuleBase
+
+__all__ = [
+    "PositionWiseFF",
+    "RelativeMultiheadAttention",
+    "RelativePositionalEncoding",
+]
+
+
+class PositionWiseFF(ModuleBase):
+
+    def __init__(self, hparams=None):
+        super().__init__(hparams)
+
+        hidden_dim = self._hparams.hidden_dim
+        ffn_inner_dim = self._hparams.ffn_inner_dim
+        dropout = self._hparams.dropout
+        activation = self._hparams.activation.capitalize()
+        if activation == 'Relu':
+            activation = 'ReLU'
+        elif activation == 'Gelu':
+            activation = 'GPTGELU'
+
+        self.linear1 = nn.Linear(hidden_dim, ffn_inner_dim)
+        self.activation_fn = get_layer({"type": activation})
+        self.dropout = nn.Dropout(dropout, inplace=True)
+        self.linear2 = nn.Linear(ffn_inner_dim, hidden_dim)
+        self.layer_norm = nn.LayerNorm(hidden_dim, eps=1e-12)
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
+        return {
+            "hidden_dim": 768,
+            "ffn_inner_dim": 3072,
+            "dropout": 0.1,
+            "activation": 'relu',
+        }
+
+    def forward(self,  # type: ignore
+                input: torch.Tensor) -> torch.Tensor:
+        # position-wise feed-forward
+        output = self.linear1(input)
+        output = self.activation_fn(output)
+        output = self.dropout(output)
+        output = self.linear2(output)
+        output = self.dropout(output)
+        # residual + layer norm
+        output = self.layer_norm(input + output)
+        return output
+
+
+class PositionalEmbedding(nn.Module):
```

Comment:
`texar.modules.embeddings` already has Position Embedders. If this is a different embedder, pls make a new Embedder class in `texar.modules.embeddings`, following the conventions of Position Embedders as much as possible

Reply:
This is almost the exact positional embedder as the one we have in Texar, but they're slightly different. Texar uses `num_timescales - 1` as denominator in the exponent while here it is just `num_timescales`. I'm wondering what was the intention of using `num_timescales - 1`, because I think (and correct me if I'm wrong) the implementation here is the orthodox version.

Reply:
`SinusoidsPositionEmbedder` is by @haoransh, can you confirm with him?

Reply:
It seems to be because that was how `tensor2tensor` had implemented it. I would prefer changing our code to match the paper description. This shouldn't affect anything because the only other place we used `SinusoidsPositionalEmbedder` is in the seq2seq Transformer example, where we don't need to load existing models.

Reply:
Would this affect the transformer performance by any chance? Do you think it's necessary to have an option in `hparams` so that users can choose `num_timescales - 1` if it's used for transformer?

Anyway, this embedder should be deleted and use the `SinusoidsPositionEmbedder` directly

Reply:
I don't really think so, as long as the model is trained from scratch.

I think the implementation here (which I adapted from the TransformerXL repo) is much simpler (and possibly faster) so maybe we can change `SinusoidsPositionEmbedder` to this implementation.